astropy/table/column.py

```diff
@@ -790,8 +793,20 @@ def __bytes__(self):
         __str__ = __bytes__
 
     def _check_string_truncate(self, value):
+        """
+        Emit a warning if any elements of ``value`` will be truncated when
+        ``value`` is assigned to self.
+        """
+        # Convert input ``value`` to the string dtype of this column
         value = np.asanyarray(value, dtype=self.dtype.type)
-        if value.dtype.itemsize > self.dtype.itemsize:
+
+        # Compute itemsize for ``value`` as the max *actual length* times the
+        # itemsize for a length=1 version of the string type.  Remember itemsize
+        # is the total length in bytes for a single item.
+        value_str_len = np.char.str_len(value).max()
+        value_itemsize = value_str_len * np.array('', dtype=value.dtype.type).itemsize
```

Comment:
It is bizarre that the number of elements cannot be grabbed more directly!

In any case, in the current version, I'd make the string `'1'`, just in case zero-length strings become possible (which is planned; https://github.com/numpy/numpy/pull/8970).

Alternatively, I guess one could do
```
self_max_len = int(self.dtype.str[2:])
if value_str_len > self_max_len:
```